lua/entities/gmod_tardis/modules/sh_destination.lua

```diff
@@ -92,22 +100,83 @@ TARDIS:AddKeyBind("destination-demat",{
     clientonly=true,
     exterior=true
 })
+TARDIS:AddKeyBind("destination-jump",{
+    name="Jump",
+    section="Destination",
+    func=function(self,down,ply)
+        if ply:GetTardisData("destination") then
+            if down then
+                self:SetData("destination-trace",true)
+            elseif self:GetData("destination-trace") then
+                local prop = self:GetData("destinationprop")
+                local pos,ang = self:GetDestinationPropTrace(ply,ply:EyeAngles())
+
+                prop:SetPos(pos)
+                prop:SetAngles(ang)
+
+                self:SetData("destination-trace",false)
+            end
+        end
+    end,
+    key=MOUSE_RIGHT,
+    clientonly=true,
+    exterior=true
+})
 TARDIS:AddKeyBind("destination-snaptofloor",{
     name="SnapToFloor",
     section="Destination",
     func=function(self,down,ply)
-        if ply:GetTardisData("destination") then
+        if down and ply:GetTardisData("destination") then
             local prop = self:GetData("destinationprop")
             if IsValid(prop) then
-                local pos = util.QuickTrace(prop:GetPos()+Vector(0,0,50), Vector(0,0,-1)*99999999).HitPos
+                local prevpos = self:GetData("destination_snaptofloor_lastpos")
+                local should_glue = (prevpos == prop:GetPos())
+
+                local pos, ang = self:GetGroundedPos(prop:GetPos(), should_glue)
                 prop:SetPos(pos)
+                prop:SetAngles(ang)
+                self:SetData("destination_snaptofloor_lastpos", (not should_glue and pos))
             end
         end
     end,
     key=KEY_R,
     clientonly=true,
     exterior=true
 })
+TARDIS:AddKeyBind("destination-rotation-toggle",{
+    name="ToggleRotationMode",
+    section="Destination",
+    func=function(self,down,ply)
+        if down and ply:GetTardisData("destination") then
+            local rotating = not self:GetData("destination-rotating")
+            self:SetData("destination-rotating", rotating)
+            TARDIS:StatusMessage(ply, "Destination.RotationMode", rotating)
+        end
+    end,
+    key=KEY_T,
+    clientonly=true,
+    exterior=true
+})
+TARDIS:AddKeyBind("destination-find-random",{
+    name="FindRandom",
+    section="Destination",
+    func=function(self,down,ply)
+        if down and ply:GetTardisData("destination") then
+            local prop = self:GetData("destinationprop")
+            if IsValid(prop) then
+                prop:SetAngles(Angle(0,prop:GetAngles().y,0))
+                local grounded = not self:GetData("destination_random_grounded")
+                self:SetData("destination_random_grounded", grounded)
```

Comment:
Is this intended to flip through a sky pos and ground pos on every press? Feels a little weird

Reply:
Random destination can be found in the sky or on the ground
That's why I alternate them
I may just leave the grounded version here if you want 
(or make a condition for the sky pos like `shift`/`boost` pressed with G - what do u think?)

Reply:
I think a condition would make sense yes!

Reply:
Made it always stick to the ground now